txscript/script.go

```diff
@@ -731,6 +731,66 @@ func GetPreciseSigOpCount(scriptSig, scriptPubKey []byte, bip16 bool) int {
 	return getSigOpCount(shPops, true)
 }
 
+// GetWitnessSigOpCount returns the number of signature operations generated by
+// spending the passed pkScript with the specified witness, and or sigScript.
+// Unlike GetPreciseSigOpCount, this function is able to accurately count the
+// number of signature operations generated by spending witness programs, and
+// nested p2sh witness programs. If the script fails to parse, then the count
+// up to the point of failure is returned.
+func GetWitnessSigOpCount(sigScript, pkScript []byte, witness wire.TxWitness) int {
+	// If this is a regular witness program, then we can proceed directly
+	// to counting its signature operations without any further processing.
+	if IsWitnessProgram(pkScript) {
+		return getWitnessSigOps(pkScript, witness)
+	}
+
+	// Next, we'll check the sigScript to see if this is a nested p2sh
+	// witness program. This is a case wherein the sigScript is actually a
+	// datapush of a p2wsh witness program.
+	sigPops, err := parseScript(sigScript)
+	if err != nil {
+		return 0
+	}
+	if IsPayToScriptHash(pkScript) &&
+		isPushOnly(sigPops) &&
+		IsPayToWitnessScriptHash(sigScript[1:]) {
+		return getWitnessSigOps(sigScript[1:], witness)
+	}
+
+	return 0
+}
+
+// getWitnessSigOps returns the number of signature operations generated by
+// spending the passed witness program wit the passed witness. The exact
+// signature counting heuristic is modified by the version of the passed
+// witness program. If the version of the witness program is unable to be
+// extracted, then 0 is returned for the sig op count.
+func getWitnessSigOps(witnessProgram []byte, witness [][]byte) int {
+	// Attempt to extract the witness program version.
+	witnessVersion, _, err := ExtractWitnessProgramInfo(witnessProgram)
+	if err != nil {
+		return 0
+	}
+
+	switch witnessVersion {
+	case 0:
+		switch {
+		case len(witnessProgram) == 22:
+			return 1
+		case len(witnessProgram) == 34 && len(witness) > 0:
+			pkScript := witness[len(witness)-1]
+			pops, err := parseScript(pkScript)
+			if err != nil {
+				return 0
```

Comment:
This is not correct and would break consensus.  It's supposed to return the count up to the point of failure when a script doesn't fully parse.  Assuming you overwrite `witnessProgram` with the extracted program as I suggested above, it should be the following:

```Go
		case len(witnessProgram) == 32 && len(witness) > 0:
			// Don't check error since parseScript returns the
			// parsed-up-to-error list of pops.
			pkScript := witness[len(witness)-1]
			pops, _ := parseScript(pkScript)
			return getSigOpCount(pops, true)
```

I'm surprised this was not caught be any tests.  There should definitely be a test added for it.